gen/field.go

```diff
@@ -461,6 +465,46 @@ func (f fieldGroupGenerator) Equals(g Generator) error {
 		`, f)
 }
 
+func (f fieldGroupGenerator) Zap(g Generator) error {
+	return g.DeclareFromTemplate(
+		`
+		<$zapcore := import "go.uber.org/zap/zapcore">
+		<$v := newVar "v">
+		<$enc := newVar "enc">
+
+		// MarshalLogObject implements zapcore.ObjectMarshaler, enabling
+		// fast logging of <.Name>.
+		func (<$v> *<.Name>) MarshalLogObject(<$enc> <$zapcore>.ObjectEncoder) error {
+			<range .Fields>
+				<- $fname := goName . ->
+				<- $fval := printf "%s.%s" $v $fname ->
+				<- if .Required ->
+					<- $encAdd := printf "%s.Add%s(%q, %s)" $enc (zapEncoder .Type) .Name (zapMarshaler .Type $fval) ->
+					<if (zapCanError .Type) ->
+						if err := <$encAdd>; err != nil {
+							return err
+						}
+					<- else ->
+						<$encAdd>
+					<- end ->
+				<- else ->
+					<- $encAdd := printf "%s.Add%s(%q, %s)" $enc (zapEncoder .Type) .Name (zapMarshalerPtr .Type $fval) ->
```

Comment:
I think we can reduce some of the duplication here with a TemplateFunc:

```
TemplateFunc("fieldMarshaler", func(required bool, t compile.TypeSpec, fval string) (string, error) {
    if required {
        return zapMarshaler(g, t, fval)
    }
    return zapMarshalerPtr(g, t, fval)
})
```

Then the template becomes something like,

```
<- $encAdd := printf "%s.Add%s(%q, %s)" ... (fieldMarshaler .Required .Type $fval) ->
<- if .Required ->
    <if (zapCanError .Type) ->
        ...
    <- else ->
        <$encAdd>
    <- end _>
<- else ->
    ...
<end>
```